x-wing/src/lib.rs

```diff
@@ -0,0 +1,406 @@
+use sha3::digest::{ExtendableOutput, XofReader};
+use sha3::{Sha3_256, Shake128, Shake128ReaderCore};
+use x25519_dalek::{x25519, X25519_BASEPOINT_BYTES};
+#[cfg(feature = "zeroize")]
+use zeroize::{Zeroize, ZeroizeOnDrop};
+
+type MlKem768DecapsulationKey = kem::DecapsulationKey<MlKem768Params>;
+type MlKem768EncapsulationKey = kem::EncapsulationKey<MlKem768Params>;
+
+const X_WING_LABEL: &[u8; 6] = br"\.//^\";
+
+/// Size in bytes of the `EncapsulationKey`.
+pub const ENCAPSULATION_KEY_SIZE: usize = 1216;
+/// Size in bytes of the `DecapsulationKey`.
+pub const DECAPSULATION_KEY_SIZE: usize = 32;
+/// Size in bytes of the `Ciphertext`.
+pub const CIPHERTEXT_SIZE: usize = 1120;
+
+/// Shared secret key.
+pub type SharedSecret = [u8; 32];
+
+// The naming convention of variables matches the RFC.
+// ss -> Shared Secret
+// ct -> Cipher Text
+// ek -> Ephemeral Key
+// pk -> Public Key
+// sk -> Secret Key
+// Postfixes:
+// _m -> ML-Kem related key
+// _x -> x25519 related key
+
+/// X-Wing encapsulation or public key.
+#[derive(Clone, PartialEq)]
+pub struct EncapsulationKey {
+    pk_m: MlKem768EncapsulationKey,
+    pk_x: x25519_dalek::PublicKey,
+}
+
+impl Encapsulate<Ciphertext, SharedSecret> for EncapsulationKey {
+    type Error = ();
+
+    fn encapsulate(
+        &self,
+        rng: &mut impl CryptoRngCore,
+    ) -> Result<(Ciphertext, SharedSecret), Self::Error> {
+        // Swapped order of operations compared to RFC, so that usage of the rng matches the RFC
+        let (ct_m, ss_m) = self.pk_m.encapsulate(rng)?;
+
+        let ek_x: SharedSecret = generate(rng);
+        let ct_x = x25519(ek_x, X25519_BASEPOINT_BYTES);
+        let ss_x = x25519(ek_x, self.pk_x.to_bytes());
+
+        let ss = combiner(&ss_m, &ss_x, &ct_x, &self.pk_x);
+
+        #[cfg(feature = "zeroize")]
+        {
+            let mut ss_x = ss_x;
+            ss_x.zeroize();
+        }
+
+        let ct = Ciphertext { ct_m, ct_x };
+        Ok((ct, ss))
+    }
+}
+
+impl EncapsulationKey {
+    /// Convert the key to the following format:
+    /// ML-KEM-768 public key(1184 bytes) | X25519 public key(32 bytes).
+    #[must_use]
+    pub fn as_bytes(&self) -> [u8; ENCAPSULATION_KEY_SIZE] {
+        let mut buffer = [0u8; ENCAPSULATION_KEY_SIZE];
+        buffer[0..1184].copy_from_slice(&self.pk_m.as_bytes());
+        buffer[1184..1216].copy_from_slice(self.pk_x.as_bytes());
+        buffer
+    }
+}
+
+impl From<&[u8; ENCAPSULATION_KEY_SIZE]> for EncapsulationKey {
+    fn from(value: &[u8; ENCAPSULATION_KEY_SIZE]) -> Self {
+        let mut pk_m = [0; 1184];
+        pk_m.copy_from_slice(&value[0..1184]);
+        let pk_m = MlKem768EncapsulationKey::from_bytes(&pk_m.into());
+
+        let mut pk_x = [0; 32];
+        pk_x.copy_from_slice(&value[1184..]);
+        let pk_x = x25519_dalek::PublicKey::from(pk_x);
+        EncapsulationKey { pk_m, pk_x }
+    }
+}
+
+/// X-Wing decapsulation key or private key.
+#[derive(Clone, PartialEq, Eq)]
```

Comment:
Derived `PartialEq`/`Eq` can leak information about the private key via short-circuiting comparisons.